src/Fantomas.Core/CodePrinter.fs

```diff
@@ -2658,7 +2658,16 @@ and genMultiLineArrayOrListAlignBrackets (isArray: bool) xs openingTokenRange cl
         +> indent
         +> sepNlnUnlessLastEventIsNewline
         +> col sepNln xs (genExpr astContext)
-        +> unindent
+        +> (fun (ctx: Context) ->
+            match Queue.tryHead ctx.WriterEvents with
```

Comment:
The last `expr` in `xs` had a trivia after in this case.
Leading the last event to be a `WriteLineBecauseOfTrivia`.
This of course messes with the `unindent` call later on, as indentation changes only kick in on the next line.

In this function, I reset the last event and play it again, where I'm adding `unindent` before the  `sepNlnForTrivia`.

Is this in principle somewhat ok to do? If so, we probably want to extract a helper in `Context`.

Reply:
> Is this in principle somewhat ok to do @jindraivanek? If so, we probably want to extract a helper in `Context`.

In general, not exactly, there is more to it than changing `Lines`, see https://github.com/fsprojects/fantomas/blob/master/src/Fantomas.Core/Context.fs#L83 Maybe it doesn't matter in this case, but not sure about that.

Correct way to do it would be to replay changed WriterEvents from some context before `WriteLineBecauseOfTrivia` is added. (As we can't easily reverse events.) I can later describe in more details of what I have in mind, if needed.

Reply:
I guess the main bit that could be off in the current approach is `let currentLine = String.Concat(List.head m.Lines, m.WriteBeforeNewline).TrimEnd()`.

Though getting the context before `WriteLineBecauseOfTrivia` is difficult as it will be in the last `genExpr astContext`. Please elaborate on the `more details` bit, I might need a hint for this one.

Reply:
Something like this:

```fsharp
        genTriviaFor SynExpr_ArrayOrList_OpeningDelimiter openingTokenRange sepOpenLFixed
        +> indent
        +> sepNlnUnlessLastEventIsNewline
        +> (fun (ctx: Context) ->
            let ctx2 = (col sepNln xs (genExpr astContext)) ctx
            let newEvs = ctx2.WriterEvents |> Seq.take (ctx2.WriterEvents.Length - ctx.WriterEvents.Length)
            match Seq.tryHead newEvs with
            | Some WriterEvent.WriteLineBecauseOfTrivia ->
                // apply new events on ctx, except the last one
                let resetModel = 
                    (ctx.WriterModel, Seq.rev(Seq.tail newEvs))
                    ||> Seq.fold (fun m e -> WriterModel.update ctx.Config.MaxLineLength e m)
                let resetContext =
                    { ctx2 with
                        WriterEvents = ctx2.WriterEvents.Tail
                        WriterModel = resetModel }
                (unindent +> sepNlnForTrivia) resetContext
            | _ -> unindent ctx2)
        +> sepNlnUnlessLastEventIsNewline
        +> genTriviaFor SynExpr_ArrayOrList_ClosingDelimiter closingTokenRange sepCloseLFixed
```

this code probably don't work (didn't try it), but hopefully main idea is self-evident.

Reply:
Oh, I see now what you mean. Good thinking, this will indeed be the safest thing to do.